### DISCUSSION TITLE
Adding a vulnerabilities table column in PR bodies

### DISCUSSION THREAD

**Discussion by or-adar** [renovatebot/renovate#14044] [Request Help] [answered]
### What would you like Renovate to be able to do?

Adding a vulnerabilities table column in PR bodies

### If you have any ideas on how this should be implemented, please tell us here.

Hi I would like to see a vulnerabilities coloumn in renovate's PR bodies.
I have some ideas of how to acheive what I'm looking for, but can't really figure out how I could put this all together, and I think that this need some tweaks on your side.

So, Basically renovate could use some open APIs to acheive what I want to get from renovate on a given PR.
For an example, let's say that I'm using npm as my package manager and I have a dependency of lodash. on a given PR, renovate suggested to move lodash to version 4.17.19;
in that case, I would like to have a column on the PR table which includes the vulnerabilites count with Synk's open API.
renovate could call [https://snyk.io/test/npm/lodash/4.17.19?type=json](https://snyk.io/test/npm/lodash/4.17.19?type=json) which will than return the following response:
(please note that only the `totalVulns` field interests me)
```json
{"resultTitle":"lodash@4.17.19","totalVulns":3}
```
It could also be really cool if I could click on that custom column and get redirected to a page on Synk (that is open to all) which lists it's vulnerabilities, by using the following url:
[https://snyk.io/vuln/npm:lodash@4.17.19](https://snyk.io/vuln/npm:lodash@4.17.19).

To sum it all up, here's a working demo of what I wish to see:
[![Vulnerability Badge](https://img.shields.io/badge/vulnerabilities-3-yellow)](https://snyk.io/vuln/npm:lodash@4.17.19)
You can click on this badge and get redirected to a vulnerabilities page on Synk

(note that the badge's `<value>` is hardcoded as 3
and the badge is generate with the following API:
`https://img.shields.io/badge/<label>-<value>-<color>`)
Here's the source of the badge's markdown:
```
[![Vulnerability Badge](https://img.shields.io/badge/vulnerabilities-3-yellow)](https://snyk.io/vuln/npm:lodash@4.17.19)
```

I though of maybe acheiving this with adding `prBodyDefinitions` to my `renovate.json` file, but this requires some programmatic tweaks (as I need to retreive `totalVulns` from the response and embed the parsed value to my custom column's template),
and I don't know if this is supported by the markdown language.
If there's any way of acheiving this without any support on your side (by using `prBodyDefinitions` or somethig else) please let me know!

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment, marked as answer]
You can already customize columns like that today in config using a combination of `prBodyDefinitions` and `prBodyColumns`. You can look at the way we added Merge Confidence badges similarly: https://github.com/whitesource/merge-confidence/blob/main/beta.json

**Reply by or-adar** [reply]
Hi @rarkins, appreciate your quick response!
Yeah, as I've said right at the ending of my issue, I was thinking about using `prBodyDefinitions`,
but there is something that isn't clear from the docs, as I'm doing several steps to generate an appropriate badge.
For instance, for the first step I'll need to call 
`https://snyk.io/test/{{datasource}}/{{replace '/' '%2f' depName}}/{{newVersion}}?type=json`
But this will return a response in the form of JSON, and what I'm trying to figure out is, if there's any way to manipulate and embed data from that request (as it is not an image, but a json response)

**Reply by rarkins** [reply]
Oh, sorry - I thought you were using an existing badge. Please see https://github.com/badges/shields/tree/master/services/snyk for what exists already

**Reply by or-adar** [reply]
@rarkins 
That's fine, I've actually found out right now that this endpoint also produces a badge (with the right parameters), like so:
`https://snyk.io/test/npm/lodash/4.17.19/badge.svg` (concating /bade.svg instead of ?type=json)
so this could get embeded like this: ![Synk Badge](https://snyk.io/test/npm/lodash/4.17.19/badge.svg)
and now I could actually use this like the Merge Confidence badge.

**Reply by LeoniePhiline** [reply]
@or-adar Would you mind sharing the config you used to add the badges?

**Reply by or-adar** [reply]
@LeoniePhiline Hi, I haven't actually configured the badge yet, but my intention is to make the following config file:
```json
{
    "prBodyDefinitions": {
            "Known Vulnerabilities": "[![Vulnerabilities of \"{{depName}}@{{newVersion}}\"](https://snyk.io/test/{{datasource}}/{{replace '/' '%2f' depName}}/{{newVersion}}/badge.svg)](https://snyk.io/test/{{datasource}}/{{replace '/' '%2f' depName}}/{{newVersion}})"
    },
    "prBodyColumns": [
        "Package",
        "Update",
        "Change",
        "Age",
        "Adoption",
        "Passing",
        "Known Vulnerabilities",
        "Confidence"
    ]
}
```
Haven't tested it, so can't tell if there's some syntax errors. but I'm pretty confident about it. you can try that anyways yourself, and I'll try to update you when I'll add it myself.

BTW, @rarkins do I always have to specify `prBodyColumns` when using `prBodyDefinitions` to add a column? even if I extend the base config which includes by default "Package", "Age", "Adoption", "Passing" etc..
is there a way to only specify what I've added? (which is the `"Known Vulnerabilities"` column)

**Reply by rarkins** [reply]
You must specify all. And generally shorter titles look better

**Reply by or-adar** [reply]
@rarkins Got it! Thanks!
